cpp/src/parquet/encoding.cc

```diff
@@ -1854,15 +1920,30 @@ void DictDecoderImpl<ByteArrayType>::InsertDictionary(::arrow::ArrayBuilder* bui
   PARQUET_THROW_NOT_OK(binary_builder->InsertMemoValues(*arr));
 }
 
-class DictByteArrayDecoderImpl : public DictDecoderImpl<ByteArrayType>,
-                                 virtual public ByteArrayDecoder {
+template <>
+void DictDecoderImpl<LargeByteArrayType>::InsertDictionary(
+    ::arrow::ArrayBuilder* builder) {
+  auto binary_builder = checked_cast<::arrow::LargeBinaryDictionary32Builder*>(builder);
+
+  // Make a LargeBinaryArray referencing the internal dictionary data
+  auto arr = std::make_shared<::arrow::LargeBinaryArray>(
+      dictionary_length_, byte_array_offsets_, byte_array_data_);
```

Comment:
Maybe you can:
1. Casting `int32_t` to `int64_t` all-at-once when neccessary
2. define a trait like `offset_type`, and use correspond `byte_array_offsets_`.

Reply:
Hm.. This might actually be a problem if I understood it correctly. ::arrow::LargeBinaryArray uses ::arrow::LargeBinaryType, which defines offset_type to be 64 bits. The byte_array_offsets_ object is a buffer, so I assume that upon reading there'll be a "blind" cast that assumes offsets are 64 bit long, but since we are passing a buffer with 32 bit offsets, it'll be a problem.

Is that correct?

Reply:
Yes, you can take a look at `BaseBinaryArray` and constructor for `LargeBinaryArray`. Seems now we regard it as `int32_t` when allocating, setting and building

Reply:
`byte_array_offsets_` is a `ResizableBuffer`, so we have the chance to allocate it as int64_t for LargeByteArrayType

Reply:
`byte_array_offsets_` is a `ResizableBuffer`, so we have the chance to allocate it as int64_t for LargeByteArrayType

Reply:
Hm.. Not sure I follow, you mean allocate it as a `int64_t` on the below code?

```
    PARQUET_THROW_NOT_OK(
        byte_array_offsets_->Resize((dictionary_length_ + 1) * sizeof(int32_t),
                                    /*shrink_to_fit=*/false));
```

That does not seem to make much sense to me, because it stores `ByteArray` (contains 32bit length). Plus, the logic on `DictDecoderImpl::SetByteArrayDict` would have to change. I might also be confusing it with the dictionary we agreed would be 32 bit

Reply:
> That does not seem to make much sense to me, because it stores ByteArray (contains 32bit length).

I guess `byte_array_offsets_` stores `ResizeableBuffer` and regard them as `int32_t` array. `byte_array_data_` is for `ByteArray`.

`int32_t` is enough for length, but here we just wants an offset

Reply:
hm, that's true, my bad for confusing `byte_array_data_` with `byte_array_offsets_`. 

In this case, it looks like defining an `offset_type` trait might do the trick and we wouldn't need to modify the offset_type during "calculation", it can still be of type `int32_t` since we are not expecting to go above that. I am referring to the below code as "offset calculation":

 ```
   int32_t offset = 0;
    uint8_t* bytes_data = byte_array_data_->mutable_data();
    int32_t* bytes_offsets =
        reinterpret_cast<int32_t*>(byte_array_offsets_->mutable_data());
```

I believe we are only interested in modifying the storage type / size and on `bytes_offsets[i] = offset` the implicit cast will do the trick.

Please correct me if I am wrong

Reply:
I'm not sure I understand what you mean.

> we wouldn't need to modify the offset_type during "calculation"

Maybe. I guess the actual code would better be something like:

1.

```c++
offset_t offset = 0;
   uint8_t* bytes_data = byte_array_data_->mutable_data();
   offset_t* bytes_offsets =
       reinterpret_cast<offset_t*>(byte_array_offsets_->mutable_data());
```

In this case, `byte_array_offsets_` are allocated with length `(size + 1) * sizeof(offset_t)`, and using `offset_t` to write. The underlying data is int32_t, so maybe it includes:

```
bytes_offsets[idx] (which could an int64_t or int32_t) = len + bytes_array[idx].len; (len should be int32_t) 
```

And when building, just use `byte_array_offsets_`

2.

Not change the `byte_array_offsets_`. And when building, casting it to a int64_t array with length `(size + 1) * sizeof(offset_t)`


Personally I perfer (1).

Reply:
Sorry if it wasn't clear, I meant (1). Will see if I can implement it then!